src/Simulation/Simulators/ToffoliSimulator/Processor.cs

```diff
@@ -0,0 +1,518 @@
+                {
+                    borrowedQubitStates.Remove(qubit);
+                }
+            }
+        }
+
+        // An X operation inverts the bit in the `simulationValues` variable at
+        // the position of the qubit's index.
+        public override void X(Qubit qubit)
+        {
+            InvertValue(qubit);
+        }
+
+        // If the simulation values of all control qubits are assigned true,
+        // the simulation value of the target qubit is inverted.
+        public override void ControlledX(IQArray<Qubit> controls, Qubit qubit)
+        {
+            if (controls.All(control => GetValue(control)))
+            {
+                InvertValue(qubit);
+            }
+        }
+
+        // SWAP interchanges the simulation values at two locations.
+        public override void SWAP(Qubit qubit1, Qubit qubit2)
+        {
+            var tmp = GetValue(qubit1);
+            SetValue(qubit1, GetValue(qubit2));
+            SetValue(qubit2, tmp);
+        }
+
+        public override void ControlledSWAP(IQArray<Qubit> controls, Qubit qubit1, Qubit qubit2)
+        {
+            if (controls.All(control => GetValue(control)))
+            {
+                var tmp = GetValue(qubit1);
+                SetValue(qubit1, GetValue(qubit2));
+                SetValue(qubit2, tmp);
+            }
+        }
+
+        /// <summary>
+        /// Make sure that a rotation is either the identity or an X, up to global phase.
+        /// Assumes that the rotation is exp(i*theta*sigma(axis)), where sigma(axis) is
+        /// the appropriate Pauli matrix normalized to +/-1 eigenvalues.
+        /// Throws an exception if the rotation is invalid.
+        /// </summary>
+        /// <param name="axis">The rotation axis</param>
+        /// <param name="theta">The rotation angle, properly normalized</param>
+        /// <returns>A pair of flags. The first indicates whether the rotation is
+        /// equivalent to X; the second whether or not the rotation is exactly equal
+        /// to the identity (no phase), and thus may be controlled.</returns>
+        private (bool isX, bool safe) CheckRotation(Pauli axis, double theta)
+        {
+            // Rotations around X of odd multiples of pi/2 are X (up to global phase).
+            // Rotations around any axis of multiples of pi are I (up to global phase).
+            // Rotations of any angle around I are I.
+            // Rotations around any axis of even multiples of pi are safe to control.
+
+            var piOverTwoRatio = Abs(theta) / (PI / 2.0);
+            if (Abs(piOverTwoRatio - Round(piOverTwoRatio)) > Tolerance)
+            {
+                throw new InvalidOperationException($"The Toffoli simulator can only perform rotations of multiples of pi/2.");
+            }
+
+            var piOverTwoCount = (int)Round(piOverTwoRatio);
+
+            var equivalentToX = (axis == Pauli.PauliX) && (piOverTwoCount % 2 == 1);
```

Comment:
>var [](start = 12, length = 3)

This is just a bool, right? please help me understand the code faster by stating it. Here and everywhere.